tests/plenary/async_lib/semaphore_spec.lua

```diff
@@ -12,6 +12,47 @@ a.describe('semaphore', function()
     eq(status, false)
   end)
 
-  a.it('should count properly', function()
+  a.it('should acquire a permit if available', function ()
+    local sem = a.util.Semaphore.new(1)
+    local permit = await(sem:acquire())
+    assert(permit ~= nil)
+  end)
+
+  a.it('should block if no permit is available', function ()
+    local sem = a.util.Semaphore.new(1)
+    await(sem:acquire())
+    
+    local completed = false
+    local blocking = async(function ()
+      await(sem:acquire())
+      completed = true
+    end)
+    a.run(blocking())
+
+    eq(completed, false)
+  end)
+
+  a.it('should give another permit when an acquired permit is released', function ()
+    local sem = a.util.Semaphore.new(1)
+    local permit = await(sem:acquire())
+    permit:forget()
+    local next_permit = await(sem:acquire())
+    assert(next_permit ~= nil)
+  end)
+
+  a.it('should permit the next waiting client when a permit is released', function () 
+    local sem = a.util.Semaphore.new(1)
+    local permit = await(sem:acquire())
+
+    local completed = false
+    local blocking = async(function ()
+      await(sem:acquire())
+      completed = true
+    end)
+
+    a.run(blocking())
```

Comment:
I used `a.run` this time to avoid blocking, since I need to release the permit after. `run` should at least immediately unwind the coroutine until the next yield, so the acquire definitely blocks at that point (confirmed with print logs). I'm not sure how stable that test is though; never failed after a couple runs anyway.

Any suggestion how to sequence that better?

Reply:
Actually, I believe calling forget should then unwind the rest of the coroutine synchronously since it calls the stored continuation... right?

Reply:
Yes `forget` will unwind the coroutine because it stored the callback

Reply:
The test is very stable because we are not doing any io so everything will always run the same way.